Super phantasm

unknownoperation opened the pull request:
Merge initial prototype to master

unknownoperation committed:
added export header file to the core

unknownoperation committed:
added scene classes
added vec & matr classes
added command line parser

LeoMash committed:
basic architecture fixes by unknownoperation

unknownoperation committed:
fixed math module
moved some classes to core
changed class names
optimized console parser function

unknownoperation committed:
*added math tester module
*fixed image storage class
*fixed math module

unknownoperation committed:
*minor classes refactoring
*added physical directories for classes in "phantasm-core"
*added own "main" function in "phm-core-tester"

unknownoperation committed:
* minor fixes in math module & math tester
* added "jpeg-compressor" and ability to save scene to "jpeg" image

unknownoperation committed:
* added "gtest" libs for both platforms and both configurations

unknownoperation committed:
*added "RAY" class
*added intersection of ray with scene objects
*added simple scene
*added simple raytracing

unknownoperation committed:
*added "jsoncpp"
*added loading scene from "*.json" file
*fixed bug with near distance in tracer

unknownoperation committed:
*added "material" filter and files
*fixed all warnings

LeoMash commented on phantasm/scene.json:
```diff
@@ -0,0 +1,18 @@
+{
```
неплохо бы завести спец директорию, куда перенести этот файл. не надо в корне держать

LeoMash commented on phantasm/src/phantasm-core/image_saver/image_saver.cpp:
```diff
@@ -0,0 +1,10 @@
+#include <fstream>
+
+#include "..\..\..\3rd-party\jpeg-compressor\jpge.h"
```
не надо таких инклудов. добавь в настройки проекта корневые директории для поиска и все инклуды пусть будут относительными

LeoMash commented on phantasm/src/phantasm-core/image_saver/image_saver.h:
```diff
@@ -0,0 +1,13 @@
+
+#include "..\image_storage\image_storage.h"
+
+void PHM_CORE_API SaveImageToJpeg (const IMAGE_STORAGE & img, std::string fileName);
```
ты уверен что PHM_CORE_API в правильном месте находится? мне казалось что он перед декларацией должен быть

LeoMash commented on phantasm/src/phantasm-core/math/mat.cpp:
```diff
@@ -0,0 +1,189 @@
+#include "..\common.h"
```
ну переименуй ты в matr.cpp его

LeoMash commented on phantasm/src/phantasm-core/math/vecmat.h:
```diff
@@ -0,0 +1,73 @@
+#pragma once
```
не нравится название vecmat. сделай по нормальному ашнику для вектора и для матрицы.
а потом можно сделать какой-нить общий ашник, куда наинклудить все составные

LeoMash commented on phantasm/src/phantasm-core/objects/object.h:
```diff
@@ -0,0 +1,34 @@
+#include "..\common.h"
+#include "..\material\material.h"
+
+struct INTERSECT_PARAMS {
```
уверен что это PARAMS ?

LeoMash commented on phantasm/src/phantasm-core/objects/object.cpp:
```diff
@@ -0,0 +1,23 @@
+
+}
+
+INTERSECT_PARAMS OBJECT::Intersect(const RAY & ray)
```
про чисто виртуальные функции слышал?

LeoMash commented on phantasm/src/phantasm-core/objects/sphere.h:
```diff
@@ -0,0 +1,20 @@
+   SPHERE (double rad, VEC pos, RGB color);
+   ~SPHERE (void);
+
+   INTERSECT_PARAMS Intersect (const RAY & ray);
```
используй override

LeoMash commented on phantasm/src/phantasm-core/scene/scene.h:
```diff
@@ -0,0 +1,22 @@
+#include "..\objects\object.h"
+#include "..\objects\sphere.h"
+
+struct SCENE_IMPL;
```
перенеси этот forward decl внутрь SCENE и переименуй просто в IMPL

LeoMash commented on phantasm/src/phantasm-core/scene/scene.cpp:
```diff
@@ -0,0 +1,54 @@
+
+#include "scene.h"
+
+struct SCENE_IMPL {
```
struct SCENE::IMPL {

LeoMash commented on phantasm/src/phantasm-core/scene/scene.cpp:
```diff
@@ -0,0 +1,54 @@
+
+SCENE::~SCENE (void)
+{
+   pImpl->Clear();
```
зачем делать Clear, если у тебя impl живет от конструктора до деструктора

LeoMash commented on phantasm/src/phantasm-core/scene_loader/scene_loader.cpp:
```diff
@@ -0,0 +1,42 @@
+
+   bool parsingSuccessful = reader.parse(inFile, root);
+
+   inFile.close();
```
ifstream - это RAII-шный файл, его можно не закрывать явно, он сам закроется по выходу из скоупа. этим можно и нужно пользоваться

LeoMash commented on phantasm/src/phantasm-core/tracer/ray.h:
```diff
@@ -0,0 +1,16 @@
+   RAY (void) {}
+   ~RAY (void) {}
+
+   VEC startingPnt;
```
ну тут ты переборщил с названием, неудобно будет вычисления писать

LeoMash commented on phantasm/src/phantasm-core/tracer/tracer.cpp:
```diff
@@ -0,0 +1,44 @@
+         scn.IntersectAll(ray, prms);
+         double minDist = 1000000.0;
+         std::for_each(prms.begin(), prms.end(),
+            [&](INTERSECT_PARAMS prm) { 
```
я рекомендую явно указывать в замыкании, на что ты ссылаешься, можно нежданчики получить

LeoMash commented on phantasm/src/phantasm/console/console.h:
```diff
@@ -0,0 +1,15 @@
+
+#include <string>
+
+struct CONSOLE_PARAMETERS {
```
вбей в гугле слово CONSOLE и попробуй проассоциировать тем, как ты назвал эту структуру.
ПАРАМЕТРЫ КОНСОЛИ ?!?!?

LeoMash commented on phantasm/src/phantasm/main.cpp:
```diff
@@ -1,23 +1,33 @@
+   img.w = 500;
+   img.h = 500;
+
+   img.data = new BYTE[500 * 500 * 3]();
```
ты заводил класс, чтобы в юзер коде писать вот такое?

LeoMash commented on phantasm/3rd-party/.gitignore:
```diff
@@ -0,0 +1 @@
+!*
```
не могу комментировать по файлам, а не по строкам (кстати в p4swarm это можно и это удобно)
1) в папке bin лежат exe и dll/so
в папке lib лежат lib/a
- это стандарт. так делают все.
2) разве так (GoogleTestCompilingLibrary.lib) называются либы? зачем ты их переименовал?

unknownoperation committed:
*minor fixes related to coding principles

unknownoperation committed:
*added scene iterator
*changed 3rd-party directories organization
*added logger files
*fixed strange file paths in code
*set custom working directory for "phantasm" & "phantasm-core"
*fixed mess with using json in scene loader

unknownoperation committed:
*added camera to scene
*added lightning & point lightning classes
*divided vecmat.h into vec.h & matr.h
*added filter in "phantasm-core" project for 3-rd party
*fixed increment operators in scene iterators
*added logging
*updated "RGB" class

unknownoperation committed:
*added logger levels
*added array of intersections in tracing
*added camera in scene file
*added lightning in scene file
*change working directory
*added scene iterator on templates

unknownoperation committed:
*fixed camera scaling
*changed input file format
*fixed color glutting

unknownoperation committed:
*added reflection
*added plane in scene
*added image configuration setup in console
*added new logger functions

unknownoperation committed:
*added PNG support
*fixed plane normal in scene

unknownoperation committed:
*divided Trace function to Trace and Shade functions
*made LoadSceneFromJSON function very small & removed code duplication there
*added scene depth to JSON parameters
*removed code duplication from SCENE class
*changed CAM class constructor
*fixed memory leaks in SCENE class destructor

unknownoperation committed:
*added refraction
*added new scenes
*fixed sphere intersection
*fixed shadow ray computing
*added checking for being inside of an object
*made trace & shade functions smaller

unknownoperation committed:
*added duration counting
*make test scene more replete

unknownoperation committed:
*added duration counting
*make test scene more replete
